### PR TITLE
feature(nemesis): introduce `raise_exception_in_thread`

### PR DESCRIPTION
**Pull request by fruch**
`raise_exception_in_thread` is introduced as a way to force kill a nemesis thread for the following reasons:

a) so we don't wait for 30min at the end of every test (can be
   more depend on the nemsis/testcase))
b) so nemesis are depended on the load (ex. SLA), and since we
   kill the load before stopping the nemesis, if we are currently
   during such a nemesis, it would spit false posisitve events
   since it won't find enough load on the cluster as it expects.

Fixes: #6439

### Testing
- [x] - https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/fruch/job/longevity-100gb-4h-test/25/
- [x] - https://argus.scylladb.com/test/4f234961-de7a-4d4a-a0bb-0795c66a7212/runs?additionalRuns[]=20d3d721-2bb5-4150-a423-6839d9603c44

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

File: sdcm/utils/common.py [hunk 1]
```diff
@@ -3143,3 +3143,8 @@ def SoftTimeoutContext(timeout: int, operation: str):  # pylint: disable=invalid
     if duration > timeout:
         SoftTimeoutEvent(operation=operation, soft_timeout=timeout,
                          duration=duration).publish_or_dump()
+
+
+def raise_exception_in_thread(thread: threading.Thread, exception_type: Type[BaseException]):
+    res = ctypes.pythonapi.PyThreadState_SetAsyncExc(ctypes.c_ulong(thread.ident), ctypes.py_object(exception_type))
```
Comment (juliayakovlev) [inline sdcm/utils/common.py:3149]: Maybe we should raise if result of the api call is not as expected, like:
```
res = pythonapi.PyThreadState_SetAsyncExc(c_long(thread.ident),
                                                      py_object(exception_type))
        if res == 0:
            raise ValueError("invalid thread id")
        elif res != 1:
            # "if it returns a number greater than one, you're in trouble,
            # and you should call it again with exc=NULL to revert the effect"
            pythonapi.PyThreadState_SetAsyncExc(c_long(tid), None)
            raise SystemError("PyThreadState_SetAsyncExc failed")
```

**Comment by juliayakovlev** [comment]
Why do we prefer to stop stress before nemesis?

**Comment by fruch** [comment]
> Why do we prefer to stop stress before nemesis?

the assumption is that without stress, most nemesis would finish much more faster.
we could try that as well, but I would still like a way to stop nemesis with further questions.

File: sdcm/utils/common.py (hunk 1 above)
Comment (fruch) [reply sdcm/utils/common.py:3149]: this case might have, if calling this function multiple times without holding the GIL.
I don't think it could in our case.

But I still didn't see it happening as I would expect

**Comment by soyacz** [comment]
This will not work sometimes (randomly) - because we often in our code use `except Exception` and we ignore/transform it.
So, we would need to retry on threads that didn't end several times or use `BaseException` as base class for `KillNemesis` - but I'm not sure if we should do it (it is not recommeded, but from the [other hand](https://docs.python.org/3/tutorial/errors.html#:~:text=BaseException%20is%20the%20common%20base,that%20the%20program%20should%20terminate.) it is used to mark fatal-error that usually should cause python exit)

**Comment by fruch** [comment]
> This will not work sometimes (randomly) - because we often in our code use `except Exception` and we ignore/transform it. So, we would need to retry on threads that didn't end several times or use `BaseException` as base class for `KillNemesis` - but I'm not sure if we should do it (it is not recommeded, but from the [other hand](https://docs.python.org/3/tutorial/errors.html#:~:text=BaseException%20is%20the%20common%20base,that%20the%20program%20should%20terminate.) it is used to mark fatal-error that usually should cause python exit)

it would be a good exercise to flush out those silly places were we used too broad Exception catching.
we are using Base in `CriticalTestFailure` on the main stress thread, so maybe that would work o.k. as well

File: sdcm/cluster.py [hunk 2]
```diff
@@ -4329,6 +4329,9 @@ def start_nemesis(self, interval=None, cycles_count: int = -1):
 
     @log_run_info("Stop nemesis threads on cluster")
     def stop_nemesis(self, timeout=10):
+        # avoid cyclic dependency between cluster.py and nemesis.py
+        from sdcm.nemesis import KillNemesis  # pylint: disable=import-outside-toplevel
```
Comment (vponomaryov) [inline sdcm/cluster.py:4333]: Looks like it is time to create dedicated module for exceptions.

File: sdcm/nemesis.py [hunk 3]
```diff
@@ -3970,8 +3977,8 @@ def disrupt_run_cdcstressor_tool(self):
         ks_tables_with_cdc = self.cluster.get_all_tables_with_cdc(self.target_node)
         if not ks_tables_with_cdc:
             self.log.warning("CDC is not enabled on any table. Skipping")
-            UnsupportedNemesis("CDC is not enabled on any table. Skipping")
-            return
+            raise UnsupportedNemesis("CDC is not enabled on any table. Skipping")
+
```
Comment (vponomaryov) [inline sdcm/nemesis.py:3981]: It looks like a totally unrelated to the PR idea.
This fix must be in a separate PR.

**Review by vponomaryov — CHANGES_REQUESTED** [review]
If the only reason to implement this PR is fixing the https://github.com/scylladb/scylla-cluster-tests/issues/6439
Then, I think, would be better to enhance that particular nemesis to be smart enough and track the load presence itself.

For example, it can call prometheus API and check even some details of the load and not just making an assumption about it's presence as it does now.

**Comment by fruch** [comment]
> If the only reason to implement this PR is fixing the https://github.com/scylladb/scylla-cluster-tests/issues/6439
> Then, I think, would be better to enhance that particular nemesis to be smart enough and track the load presence itself.
> 
> For example, it can call prometheus API and check even some details of the load and not just making an assumption about it's presence as it does now.

It's not one nemesis, it's set of them.
Also it's a long standing issue that we didn't implement a way to stop the nemesis immediately, but waiting for 30min to let it finish, which is not always the case, it's not the first time we had issues similar to that, or cases we get failure of the last nemesis being reported.

I don't want to make nemesis more complicated then necessary, and also be aware when the main load is stopped/ended.

**Comment by soyacz** [comment]
> it would be a good exercise to flush out those silly places were we used too broad Exception catching.

image
Having 106+ exceptions handled like this it won't be an easy task...

> we are using Base in `CriticalTestFailure` on the main stress thread, so maybe that would work o.k. as well

If so, we could try that way.


> If the only reason to implement this PR is fixing the #6439
> Then, I think, would be better to enhance that particular nemesis to be smart enough and track the load presence itself.

Anyway, we lack a way of stopping nemesis in case we no longer care of its result (and we must finish in short time to not get timeout error). I couldn't figure out any other feasible way of doing it than proposed by @fruch.

**Review by vponomaryov — COMMENTED** [review]
The new exceptions module could be in a separate commit.

**Comment by fruch** [comment]
> The new exceptions module could be in a separate commit.

moved it to a separate commit

**Review by vponomaryov — APPROVED** [review]
LGTM

**Review by soyacz — DISMISSED** [review]
LGTM
@fruch did you test if it works as expected? not killing too much?

**Comment by fruch** [comment]
> LGTM
> @fruch did you test if it works as expected? not killing too much?

I tested it a few times, but none of them was during a working nemesis

I'm still waiting for @juliayakovlev to try it on the original job that run into that issue

**Comment by fruch** [comment]
proven to be working better now, handling also stress failure that can happen during teardowns.

@juliayakovlev @soyacz @vponomaryov if you can give it a final review...

File: sdcm/exceptions.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+class NemesisSubTestFailure(Exception):
+    """ raised if nemesis got error from sub test
+    """
+
+
+class AuditLogTestFailure(Exception):
+    """ raised if nemesis got error from audit log validation
+    """
+
```
Comment (vponomaryov) [inline sdcm/exceptions.py:53]: precommit complains that here must be 2 empty lines.

File: sdcm/nemesis.py [hunk 5]
```diff
@@ -136,80 +152,12 @@
...
-
-class NoMandatoryParameter(Exception):
-    """ raised from within a nemesis execution to skip this nemesis"""
-
-
 class DefaultValue:  # pylint: disable=too-few-public-methods
     """
     This is class is intended to be used as default value for the cases when None is not applicable
     """
     ...
```
Comment (vponomaryov) [inline sdcm/nemesis.py:160]: Same here, one more empty line must be added.

**Review by vponomaryov — COMMENTED** [review]
Precommit and unit tests failed. Need to fix.

File: sdcm/nemesis.py (hunk 5 above)
Comment (fruch) [reply sdcm/nemesis.py:160]: fixed

File: sdcm/exceptions.py (hunk 4 above)
Comment (fruch) [reply sdcm/exceptions.py:53]: fixed

**Comment by fruch** [comment]
> Precommit and unit tests failed. Need to fix.

unit test was spot on, I was comparing the type of exception, and not the actual exception.

now it's works much better

**Comment by fruch** [comment]
> > Precommit and unit tests failed. Need to fix.
> 
> unit test was spot on, I was comparing the type of exception, and not the actual exception.
> 
> now it's works much better

and now that's a unit test to cover the case of KillNemesis exception

**Review by vponomaryov — APPROVED** [review]
LGTM

Why backport is set only to `2023.1`?
What about others?

**Review by soyacz — APPROVED** [review]
LGTM
